### ISSUE TITLE
Compilation issue in CamViz with Cmake 

### ISSUE THREAD

**Issue by pankhurivanjani** [JdeRobot/viz#33]
Hi!
While building the CamViz tool inside the viz/Camviz directory I am getting the errors of linking the header files.
I followed the usual procedure of 
      mkdir build 
      cd build 
      cmake ..
      make 
Errors like "glibmm.h: No such file or directory compilation terminated." are coming after the make command.
I fixed the errors in a few lines by changing the include path in the header file. But there are still errors in gtkmm linked header files. 
Also, I have changed a few lines in CMakeLists.txt for linking gtkmm package but unable to solve the errors.

Is anyone else able to compile and build this tool or facing similar issues?

Regards,

**Comment by fqez** [comment]
Hi, seems like you are lacking some dependencies (?). Make sure you have all the JdeRobot dependencies installed (debian package `jderobot-deps-dev`).

Regards!

**Comment by pankhurivanjani** [comment]
Yes..this was on top of my mind and I have installed this debian package already.
Still errors were there.

**Comment by fqez** [comment]
hmm weird... I will take a look when I have a lil bit of time. Are you trying to compile this component inside the JdeRobot/base compilation chain? Or have you made your own CMakeLists.txt since this repo doesn't have its own compilation chain?

**Comment by fqez** [comment]
Nontheless, in JdeRobot/base you have an script (named `install_deps.sh` inside `scripts` dir) which will try to install all the compilation dependencies too. Maybe the debian package is outdated, I don't think so tho; but I think it may be worth giving it a shot. 

Regards :)

**Comment by pankhurivanjani** [comment]
I made a few changes in this CMakeLists.txt [link](https://github.com/JdeRobot/viz/blob/master/camViz/CMakeLists.txt) to compile. 
Oh.. Then there must be some missing components in this file!

**Comment by pankhurivanjani** [comment]
Hi @fqez 
I will be working on viz tool now, especially camViz this week.
I am using ubuntu 18.04 and ROS Melodic.

1.  I installed jderobot base from here [ros-melodic-jderobot](https://github.com/JdeRobot/base/tree/ubuntu_bionic)

2. I followed binary installation instead of source one. Is it the right thing to do for melodic branch?

3. I cloned this viz repo, created build folder and tried building it with cmake
   
    cmake...

pankhuri@pankhuri-LENOVO-IDEAPAD-500-15ISK:~/viz/camViz/build$ cmake ..
-- Configuring done
-- Generating done
-- Build files have been written to: /home/pankhuri/viz/camViz/build
 
   make

pankhuri@pankhuri-LENOVO-IDEAPAD-500-15ISK:~/viz/camViz/build$ make 
[ 33%] Building CXX object CMakeFiles/camViz.dir/camViz.cpp.o
/home/pankhuri/viz/camViz/camViz.cpp:25:10: fatal error: visionlib/colorspaces/colorspacesmm.h: No such file or directory
 #include <visionlib/colorspaces/colorspacesmm.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
CMakeFiles/camViz.dir/build.make:62: recipe for target 'CMakeFiles/camViz.dir/camViz.cpp.o' failed
make[2]: *** [CMakeFiles/camViz.dir/camViz.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/camViz.dir/all' failed
make[1]: *** [CMakeFiles/camViz.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2

Since, we discussed about compilation chain thing a month ago. I would like to pick it from here. Currently I want to work with camViz can you guide me the workaround for camViz in ros-melodic?

Thank you,

**Comment by fqez** [comment]
Hi @pankhurivanjani it's nice to have you back here :)

About what you mention you are absolutely right. Nowadays the Viz repository is not working along the base repository because of the compilation chain. In order to make it work, you will need to do the following:

1. Clone base and Viz repo (if you haven't before)
```
git clone https://github.com/jderobot/base
git clone https://github.com/jderobot/viz
```

2. In the base repository, go to scripts folder and install the building dependencies **(please make sure it won't create any incompatibility with any of your installed packages. It shouldn't, but it won't harm if you check it before)**  It will take several minutes (ice takes a loooong time to install).
```
cd ~/base/scritps
sudo ./install_deps.sh
```

3. Once installed, copy the camViz directory to base
```
cp -r ~/viz/camViz ~/base/src/tools
```

4. Go to the root of base and create a build folder
```
cd ~/base
mkdir build && cd build
```
After that step you will have camViz inside the compilation chain, so it should be built along the other libraries and stuff.

5. Build the project with only camViz tool
```
cmake .. -Dbuild-default=OFF -Dbuild_camviz=ON
make -j 4
```
If everything goes OK (you will see many warnings, don't worry about those) you should have camViz tool already built.

6. Install the project in your system if you want, if not, the camViz tool executable should be in `~/base/build/src/tools/camViz`
```
sudo make install
```

Sorry for giving you a workaround instead of a proper solution to your problem. I will start with fixing this repo as soon as I have some free time for it.
If you face any kind of problem with this procedure, don't hesitate to ask for help I will be glad to help you.

Good luck!!!

**Comment by pankhurivanjani** [comment]
Hello!

Thank you for the workaround, I currently need something like this only.

'''''' Build the project with only camViz tool
cmake .. -Dbuild-default=OFF -Dbuild_camviz=ON
make -j 4
If everything goes OK (you will see many warnings, don't worry about those) you should have camViz tool already built. """

After make -j 4 command I am getting following errors:
``` [ 43%] Built target jderobotHandlers
make[2]: *** No rule to make target 'libOpenNI2.so', needed by 'src/libs/jderobotViewer/libjderobotViewer.so'.  Stop.
CMakeFiles/Makefile2:2687: recipe for target 'src/libs/jderobotViewer/CMakeFiles/jderobotViewerShare.dir/all' failed
make[1]: *** [src/libs/jderobotViewer/CMakeFiles/jderobotViewerShare.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 44%] Built target jderobotViewer
[ 48%] Built target jderobotutilshare
[ 48%] Built target jderobotcomm_py
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 

  ```

When I used just make then my errors are:

```
[ 48%] Built target comm-embedded
[ 48%] Built target jderobotcomm_py
[ 48%] Built target config_py
[ 49%] Built target depthLib
[ 50%] Built target depthLibshare
[ 51%] Built target easyiceconfig-embedded
[ 53%] Built target easyiceconfig
[ 54%] Built target test_stdutils
[ 55%] Built target demo
[ 56%] Built target test_loader
[ 56%] Built target easyiceconfig_py
[ 57%] Built target fuzzyshare
[ 58%] Built target fuzzylib
[ 67%] Built target geometry
[ 68%] Built target jderobotHandlersshare
[ 69%] Built target jderobotHandlers
make[2]: *** No rule to make target 'libOpenNI2.so', needed by 'src/libs/jderobotViewer/libjderobotViewer.so'.  Stop.
CMakeFiles/Makefile2:2687: recipe for target 'src/libs/jderobotViewer/CMakeFiles/jderobotViewerShare.dir/all' failed
make[1]: *** [src/libs/jderobotViewer/CMakeFiles/jderobotViewerShare.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2
pankhuri@pankhuri-LENOVO-
```

I think there is some dependency issue with derobotViewer here which is giving errors.

Any suggestions for this?

Thank you!

**Comment by fqez** [comment]
Hmm yeah, it seems like you are missing one of the dependencies. Try the following 

```
sudo apt install libopenni2-dev
```

and try to make a clean building this time by deleting all the cache generated files. Let me know if you face more problems. Thank you and sorry for the inconveniences :)

**Comment by pankhurivanjani** [comment]
Hi!

I had libopenni2-dev installed.
```
 sudo apt install libopenni2-dev
[sudo] password for pankhuri: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libopenni2-dev is already the newest version (2.2.0.33+dfsg-10).
0 upgraded, 0 newly installed, 0 to remove and 534 not upgraded.
 ```
Also I removed the previous installation of jderobot and installed whole thing again but I am getting the same error.

Few observations:
libOpenNI2.so is present in /usr/lib 

* Similar errors have been reported in  [Error](https://stackoverflow.com/questions/45320339/error-while-running-orb-slam2-library)
 
But the path in error is /usr/lib/... and in our error it is just 'libOpenNI2.so'

* While inspecting CMakeLIsts in [1.](https://github.com/JdeRobot/base/blob/master/CMakeLists.txt) I found these lines (line no 53-55)

INCLUDE_DIRECTORIES(/opt/jderobot/include/openni2)
LINK_DIRECTORIES(/opt/jderobot/lib/openni2)
set (CMAKE_LIBRARY_PATH /opt/jderobot/lib/openni2)

there is no openni2 folder in /opt/jderobot/lib or /include

**Comment by fqez** [comment]
Hmm this is getting weird xD

I am building the project on a clean Ubuntu system and it isn't giving me those problems. Try the following steps:

1. Clone the base repo again (I updated the script)
2.  Execute the deps installation script again (just to be sure)
3. Copy the camViz tool inside `src/tools` of the base repo
4. Install the OpenNI package `sudo apt install libopenni-dev libopenni2-dev`
5. Source ros with `source /opt/ros/melodic/setup.bash`
6. Make a build directory in the base repo and try to build it again.

Those steps work for me. Also check if after doing the `cmake .. -Dbuild-...` there is this `libOpenNI2.so` file in the build directory.

Hope it helps

**Comment by jmplaza** [comment]
Thinking about this, maybe it is time to completely refactor CamViz so it does not depend on ICE and on JdeRobot/base (and so comm library) anymore, but only in ROS1 (and ROS2) interfaces for images... It is not reasonable that camviz depends on libOpenNI2.... What do you think about it @fqez @pankhurivanjani ?

**Comment by fqez** [comment]
Yes, I think it's for the best... There is no reason for camViz to depend on the base repo anymore. And its current dependency with OpenNI2 is because how the JdeRobot compilation chain works. All of our libraries are built regardless of the tool/driver you want to compile. So some of those libs depend on OpenNI and that's why it's a required dependency. So yeah, a refactor of this (and maybe some other tools of this repo too) would be a logical step. 

**TL;DR;** I agree.

**Comment by pankhurivanjani** [comment]
Hello all!

I figured out the errors in my system, fixed them somehow.

( Attaching link for future reference [issue-solved](https://askubuntu.com/questions/616065/the-imported-target-qt5gui-references-the-file-usr-lib-x86-64-linux-gnu-li) )

With the command ``` cmake .. -Dbuild-default=OFF -Dbuild_camviz=ON ```'
 it built with no errors but there was not executable in the build path so for a temporary solution I built the whole project and got the executable for camviz for testing.

Now, I will work on refactoring camViz to support ROS2 and remove ICE. For now camViz is dependent on base for compiling due to current compilation chain as mentioned by @fqez I will see into it to get a better way.

**Comment by fqez** [comment]
Those are good news!! good to know you managed to solve it. The executable is only generated after the `make` command so it's weird that it doesn't appear in `build/src/tools/camViz` Since this is solved, I will close this issue for now. You can open another issue for the refactorization of camViz for ROS2 if you like.

Keep it up!😃
